### PR TITLE
Skip too new versioned directories of multi-release JARs when instrumenting

### LINKED ISSUE #25434: Instrumentation of multi-release JARs should not fail on newer versions
### Expected Behavior

When the multi-release JAR of the buildscript classpath contains a versioned directory for Java version that the current Gradle doesn't support, Gradle should be able to instrument the root directory and supported versioned directories without failures.

### Current Behavior (optional)

The instrumentation fails when it tries to instrument the classes in the unsupported versioned directory.

### Context

This issue makes it impossible to use newer libraries as part of the buildscript classpath, even if the library itself is backwards-compatible. It defeats the purpose of the MR-JAR.

See also #24390 and #24430.

### PR DESCRIPTION
**Pull request by mlopatkin**
ASM library that we use for bytecode instrumentation has an upper limit to the bytecode version it can understand.
When processing multi-release JARs, the restriction caused errors even when running on the supported JVM, because instrumentation was processing all classes inside the JAR regardless of whether it was multi-release or not: e.g. JAR with versioned directories for Java 21 and Java 11 was still problematic even when running on Java 11 JVM, where new version of classes compiled for Java 21 aren't actually used.

This PR rewrites the instrumentation code to skip all versioned directories which aren't supported. The resulting instrumented JARs only contain versioned directories up to latest ASM-supported Java version. This way the cached transformed JAR is independent of the version of the JVM running Gradle.

However, there's a hole left when the build runs on the not-yet-supported-by-asm (and, therefore, Gradle) JVM, as it can then load non-instrumented classes from the versioned directories of the original JARs. This is going to be addressed in some follow-up PR.

Part of #25434.

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle test this

**Comment by octylFractal** [comment]
I think this should have a test that the MRJAR entries are ignored, though I'm not sure how difficult that would be to add.

**Comment by mlopatkin** [comment]
> I think this should have a test that the MRJAR entries are ignored, though I'm not sure how difficult that would be to add.

Thanks, I've added unit test coverage. Setting up a proper integration test is harder, I don't think there is enough ROI.

**Comment by mlopatkin** [comment]
@bot-gradle test RfR

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/archive/JarTestFixture.groovy [hunk 1]
```diff
@@ -21,20 +21,34 @@ import org.apache.commons.io.IOUtils
...
+    /**
+     * Creates the fixture.
+     */
+    JarTestFixture(File file, String metadataCharset, String contentCharset, boolean checkManifest) {
          super(file, metadataCharset, contentCharset)
          this.file = file
-         isManifestPresentAndFirstEntry()
+         if (checkManifest) {
+             isManifestPresentAndFirstEntry()
+             manifest = new Manifest(IOUtils.toInputStream(content('META-INF/MANIFEST.MF'), contentCharset as String))
```
Comment (bamboo) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/archive/JarTestFixture.groovy:48]: ❓ Any particular reason not to load the manifest on demand?

**Review by bamboo — APPROVED** [review]
It's looking good overall. I made a few suggestions which could be addressed in a follow-up PR as needed.

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/archive/JarTestFixture.groovy (hunk 1 above)
Comment (mlopatkin) [reply subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/archive/JarTestFixture.groovy]: I intended to piggy-back on `isManifestPresentAndFirstEntry` initially, but discarded. Replaced with a lazy load (and made it possible to `manifest` even after creating the fixture with `checkManifest = false`).

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

**Pull request merged**